Add `deepMergeJSON` aggregate function

iskakaushik opened the pull request:
- This function will be very useful for ETL use-cases where only the diff is provided.
- This will eventually be wired up to `CoalescingMergeTree`
- It also supports `$unset: true` to mark a path as deleted

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- New Feature


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Added `deepMergeJSON` aggregate function for deep-merging JSON objects, preserving the last value per path and respecting size/path limits.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

iskakaushik committed:
Add `deepMergeJSON` aggregate function

- This function will be very useful for ETL use-cases where only the
  diff is provided.
- This will eventually be wired up to `CoalescingMergeTree`
- It also supports `$unset: true` to mark a path as deleted

iskakaushik committed:
add docs for this feature

iskakaushik committed:
fixes

iskakaushik committed:
fix style check

Avogar commented:
Unfortunately storing JSON as `map` with `Field` for each path won't work properly for some data types, because not all data types have corresponding Field type. For example:
```sql
:) select deepMergeJSON(json) as merged_json, '{"a" : "2020-01-01"}'::JSON as json, JSONAllPathsWithTypes(merged_json), JSONAllPathsWithTypes(json)

   ┌─merged_json─┬─json───────────────┬─JSONAllPaths⋯erged_json)─┬─JSONAllPathsWithTypes(json)─┐
1. │ {"a":18262} │ {"a":"2020-01-01"} │ {'a':'UInt16'}           │ {'a':'Date'}                │
   └─────────────┴────────────────────┴──────────────────────────┴─────────────────────────────┘
```

We have to store values of paths in a `String` value that contains binary encoded data type and binary serialized value (just like we store values for paths inside shared data in ColumnObject). We can use Field only for paths with type hints.

Also, I think we can add optional parameter to the aggregate function that will indicate the key that should be used for deletions, so we will use deletion logic only if this optional key is specified.

Avogar commented on src/AggregateFunctions/AggregateFunctionDeepMergeJSON.h:
```diff
@@ -0,0 +1,114 @@
+    /// Use std::map to keep paths sorted for consistent output
+    /// StringRef will point to Arena-allocated memory
+    std::map<StringRef, PathData> paths;
+    size_t row_count = 0;
```
I think this and `row_order` are useless, as they don't guarantee anything. We can just always keep the latest value added to the aggregate data. And during merging of 2 aggregate data we can treat the value fron the right data as the latest

Avogar commented on src/AggregateFunctions/AggregateFunctionDeepMergeJSON.h:
```diff
@@ -0,0 +1,114 @@
+    {
+        Field value;
+        size_t row_order;
+        // Track if this path was explicitly deleted
+        bool is_deleted = false;
```
Why do we need to store deleted paths? If it was deleted, we need to remove it or its children from the data and that's it.

Avogar commented on src/AggregateFunctions/AggregateFunctionDeepMergeJSON.h:
```diff
@@ -0,0 +1,114 @@
+constexpr size_t MAX_JSON_MERGE_TOTAL_SIZE = 100_MiB;
+
+/// JSON unset marker key
+constexpr const char * UNSET_KEY = "$unset";
```
I suggest making this key as an optional parameter of an aggregate function, so if it's not set we don't need to handle deletes at all

Avogar commented on src/AggregateFunctions/AggregateFunctionDeepMergeJSON.cpp:
```diff
@@ -0,0 +1,390 @@
+        return;
+    }
+
+    /// Check for deletion marker (backward compatibility)
+    if (isUnsetMarker(value))
+    {
+        handleDeletion(path, order, arena);
+        return;
+    }
```
What do you mean by `backward compatibility` here? How is it possible to get an `Object` inside `value`?

iskakaushik replied:
i was prototyping something earlier, i removed this.

Avogar commented on src/AggregateFunctions/AggregateFunctionDeepMergeJSON.cpp:
```diff
@@ -0,0 +1,390 @@
+    const auto & col_object = assert_cast<const ColumnObject &>(*columns[0]);
+    auto & aggregate_data = data(place);
+
+    for (size_t row = row_begin; row < row_end; ++row)
+    {
+        processColumnObject(col_object, row, aggregate_data, arena);
+        ++aggregate_data.row_count;
+    }
```
We can do it more efficiently. For typed paths we can just add last value in the column. For dynamic paths we can add last non-null value. We will need to iterate only over shared data paths.

iskakaushik replied:
Thanks for the optimization suggestion! I tried implementing it but discovered it breaks deletion markers that appear in intermediate rows. For example:

```sql
(1, '{"config": {"setting1": "value1"}}'),
(1, '{"config": {"setting1.$remove": true}}'),  -- This gets skipped
(1, '{"config": {"setting3": "value3"}}')
```

The optimization only processes the last row, missing the deletion marker in row 2. Would it be valuable to implement this optimization conditionally when no deletion key is configured? I think it will be, would it be ok if I did it in a follow-up PR?

Avogar commented on src/AggregateFunctions/AggregateFunctionDeepMergeJSON.cpp:
```diff
@@ -0,0 +1,390 @@
+        ReadBufferFromMemory buf(value_data.data, value_data.size);
+        auto type = decodeDataType(buf);
+
+        if (!isNothing(type))
```
Shared data never contains paths with NULL values

iskakaushik committed:
fix deletion parameter

iskakaushik committed:
remove the need for row number and row count

iskakaushik committed:
use ends_with

iskakaushik committed:
remove values correctly

iskakaushik committed:
default value check

alexey-milovidov committed:
Delete aspell-dict.txt (removed in master)

`aspell-dict.txt` was deleted in master along with the `aspell` and
`codespell` style checks. Removing the file here so the PR can merge
cleanly without a modify/delete conflict.
